lib/archive/unpack_unix.go

```diff
@@ -131,71 +137,104 @@ func InvokeUnpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, r
+func Unpack(op trace.Operation, tarStream io.Reader, filter *FilterSpec, root string, binPath string) (*exec.Cmd, error) {
 
 	fi, err := os.Stat(root)
 	if err != nil {
 		// the target unpack path does not exist. We should not get here.
 		op.Errorf("tar unpack target does not exist: %s", root)
-		return err
+		return nil, err
 	}
 
 	if !fi.IsDir() {
 		err := fmt.Errorf("unpack root target is not a directory: %s", root)
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	encodedFilter, err := EncodeFilterSpec(op, filter)
 	if err != nil {
 		op.Error(err)
-		return err
+		return nil, err
 	}
 
 	// Prepare to launch the binary, which will create a chroot at root and then invoke InvokeUnpack
 	// #nosec: Subprocess launching with variable. -- neither variable is user input & both are bounded inputs so this is fine
-	cmd := exec.Command("/bin/unpack", op.ID(), root, *encodedFilter)
+	// "/bin/unpack" on appliance
+	// "/.tether/unpack" inside container
+	cmd := exec.Command(binPath, op.ID(), root, *encodedFilter)
 
-	//stdin
 	stdin, err := cmd.StdinPipe()
 
 	if err != nil {
 		op.Error(err)
-		return err
+		return cmd, err
 	}
 
 	if stdin == nil {
 		err = errors.New("stdin was nil")
 		op.Error(err)
-		return err
+		return cmd, err
 	}
-	done := make(chan error)
+
 	go func() {
-		// copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
-		defer stdin.Close()
+		if tr, ok := tarStream.(*tar.Reader); ok {
+			tw := tar.NewWriter(stdin)
+			defer stdin.Close()
+			var th *tar.Header
+			for {
+				th, err = tr.Next()
+				if err == io.EOF {
+					err = nil // this just signals the end of the stream, so we don't want to pass it out to the parent process, as it doesn't signal a problem
+					tw.Close()
+					return
+				}
+				if err != nil {
+					op.Errorf("error reading tar header %s", err)
+					return
+				}
+				op.Debugf("processing tar header: asset(%s), size(%d)", th.Name, th.Size)
+				err = tw.WriteHeader(th)
+				if err != nil {
+					op.Errorf("error writing tar header %s", err)
+					return
+				}
+				switch th.Typeflag {
+				case tar.TypeReg:
+					var k int64
+					k, err = io.Copy(tw, tr)
+					op.Debugf("wrote %d bytes", k)
+					if err != nil {
+						op.Errorf("error writing file body bytes to stdin %s", err)
+						return
+					}
+				}
+			}
+		}
+		// if we're passed a stream that doesn't cast to a tar.Reader copy the tarStream to the binary via stdin; the binary will stream it to InvokeUnpack unchanged
```

Comment:
please make it two line comment

Reply:
Your editor doesn't smart-wrap long lines ;) ?